content/post/deduplicating_pga_with_apollo.md

```diff
@@ -0,0 +1,387 @@
+only identical files, the groups of vertices which share the same FuzzyWuzzy
+file name are indicated on the graph. While vertices with the same name
+often hashed to the same buckets, some didn't even end up in the same community.
+Furthermore, communities tend to group files with distinct names, e.g. look at
+the two largest communities in the graph.
+
+### Large projects 
+
+{{% caption src="/post/deduplicating_pga_with_apollo/azure.png" %}}
+Graph of the connected component with Python files described below, colored by community.
+Edges are colored depending on the vertices they link to.
+{{% /caption %}}
+
+| files count | edges count | distinct filenames count | projects count | communities count | threshold |
+|-------------|-------------|--------------------------|----------------|-------------------|-----------|
+| 4803        | 468,180     | 2954                     | 3              | 96                |  80%      | 
+
+This illustrates another feature - CCs with many files from few projects.
+Here the CC is composed of Python files stemming from
+[Microsoft Azure SDK for Python](https://github.com/Azure/azure-sdk-for-python), 
+with 23 files located in 2 other Azure projects (the CLI and the IoT SDK). 
+Given the repetitive structure of the SDK, no surprise.
+
+### Coding conventions
+
+{{% caption src="/post/deduplicating_pga_with_apollo/google.png" %}}
+Graph of the connected component with Java files described below, colored by community.
+Edges are  colored depending on the vertices they link to.
+{{% /caption %}}
+
+| files count | edges count | distinct filenames count | projects count | communities count | threshold |
+|-------------|-------------|--------------------------|----------------|-------------------|-----------|
+| 6116        | 272,016     | 3964                     | 563            | 251               |  80%      | 
+
+
+This is one of the 4 CCs which count together for 42% of all of the 
+Java files in [GoogleAds java-lib](https://github.com/googleads/googleads-java-lib).
+While the other three CCs generally correspond to the single project, our is more
+diverse. However, the files are still clustered by project and by company.
+The red community is made of the GoogleAds files (32% of all files),  
+the light blue is from [AWS Java SDK](https://github.com/aws/aws-sdk-java)
+and the yellow to the upper right is [AWS Android SDK](https://github.com/aws/aws-sdk-java) 
+(the latter two are 23% of all files).
+Green files are from [Plutext docx4j](https://github.com/plutext/docx4j) 
+(6% of all files), mixed with [eBay's developer program projects](https://github.com/eBayDeveloper/eBay_APICall_CodeSamples)
+(also 6%). The orange community is devoted to [YaviJava](https://github.com/yavijava/yavijava)
+(4.5%), and the others represent a wide range of projects from Facebook, Paypal, Apache, etc.
+There must be something deep in common for those codebases and   
```

Comment:
tofix

Reply:
which needs fixing ?

Reply:
The paragraph ends mid-sentence.

Reply:
aaaah yeah sorry thought you were talking about the links my bad

Reply:
anyway will let @vmarkovtsev complete, since he wrote this ^^

Reply:
I have a nasty feeling that I wanted to write smth smart there but somehow erased the ending :(

Reply:
This needs to be fixed before publication

Reply:
I have a nasty feeling that I fixed it once again and removed for the second time. Damn.